locopy/base.py

```diff
@@ -14,151 +14,155 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-"""Redshift Module
-Module to wrap a database adapter into a Cmd class which can be used to connect
-to Redshift, and run arbitrary code.
+"""Base Module
 """
-
 import time
 
-from .utility import validate_redshift_attributes, get_redshift_yaml
 from .logger import get_logger, DEBUG, INFO, WARN, ERROR, CRITICAL
-from .errors import (
-    RedshiftCredentialsError,
-    RedshiftConnectionError,
-    RedshiftDisconnectionError,
-    RedshiftError,
-)
+from .errors import CredentialsError, ConnectionError, DisconnectionError, DBError
+from .utility import get_redshift_yaml
 
 logger = get_logger(__name__, INFO)
 
 
-class Cmd(object):
-    """
-    Locopy class which manages connections to Redshift.  Subclassed as S3
-    elsewhere to provide the COPY functionality.
-
-    If any of host, port, dbname, user and password are not provided, a
-    config_yaml file must be provided with those parameters in.
+class Base(object):
+    """This is the base class for all DBAPI 2 database connectors which will inherit this
+    functionality. The ``Base`` class will manage connections and handle executing queries.
+    Most of the functionality should work out of the box for classes which inherit minus the
+    abstract method for ``_connect`` which may vary across databases.
 
     Parameters
     ----------
-    dbapi : DBAPI 2 module
-        A PostgreSQL database adapter which is Python DB API 2.0 compliant
+    dbapi : DBAPI 2 module, optional
+        A database adapter which is Python DB API 2.0 compliant
         (``psycopg2``, ``pg8000``, etc.)
 
-    host : str
+    host : str, optional
         Host name of the Redshift cluster to connect to.
 
-    port : int
+    port : int, optional
```

Comment:
I think we'll have to think a bit about how this should work. The `dbapi.connect` method is somewhat database-specific, so if you wanted to support different databases, you'd need to overwrite it in sub-classes.  Snowflake connector looks like `connect(user=USER, password=PASSWORD, account=ACCOUNT)`, doesn't have `host`, `port`, etc. but it does have other optional keyword params (`authenticator`, e.g.).  Details [here](https://docs.snowflake.net/manuals/user-guide/python-connector-example.html#connecting-using-the-default-authenticator), and all the params are [here](https://docs.snowflake.net/manuals/user-guide/python-connector-api.html#label-snowflake-connector-methods) (there's a bajillion).

We could just add in all the optional variables here and enforce keyword calling, or else save `self.connect_kwargs` or something here and then enforce in the sub-class which ones are needed/supported.

Reply:
Yeah, I figured they wouldn't all be the same which would mean some overriding of methods when it comes to this. The `Base` class would be more focused on 95% of cases where you would have certain things. We can msybe refactor a bit once we add snowflake. It might become a bit more obvious.